subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/rules/JacocoPluginCoverageVerificationIntegrationTest.groovy

```diff
@@ -0,0 +1,352 @@
+        succeeds TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS
+
+        then:
+        executedAndNotSkipped(TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS)
+    }
+
+    def "can ignore failures"() {
+        given:
+        buildFile << """
+            jacocoTestCoverageVerification {
+                violationRules {
+                    failOnViolation = true
+
+                    rule {
+                        $Insufficient.LINE_METRIC_COVERED_RATIO
+                    }
+                }
+            }
+        """
+
+        when:
+        succeeds TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS
+
+        then:
+        executedAndNotSkipped(TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS)
+        errorOutput.contains("Rule violated for bundle $testDirectory.name: lines covered ratio is 1.0, but expected maximum is 0.5")
+    }
+
+    @Unroll
+    def "can define same rules for multiple report tasks #tasksPaths"() {
+        given:
+        javaProjectUnderTest.writeIntegrationTestSourceFiles()
+
+        buildFile << """
+            tasks.withType(JacocoCoverageVerification) {
+                violationRules {
+                    rule {
+                        $Insufficient.LINE_METRIC_COVERED_RATIO
+                    }
+                }
+            }
+        """
+
+        when:
+        fails tasksPaths
+
+        then:
+        executedAndNotSkipped(tasksPaths)
+        errorOutput.contains("Rule violated for bundle $testDirectory.name: lines covered ratio is 1.0, but expected maximum is 0.5")
+
+        where:
+        tasksPaths << [TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS, INTEG_TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS]
+    }
+
+    @Unroll
+    def "can define different rules for multiple report tasks #tasksPaths"() {
+        given:
+        javaProjectUnderTest.writeIntegrationTestSourceFiles()
+
+        buildFile << """
+            $reportTaskName {
+                violationRules {
+                    rule {
+                        $limit
+                    }
+                }
+            }
+        """
+
+        when:
+        fails tasksPaths
+
+        then:
+        executedAndNotSkipped(tasksPaths)
+        errorOutput.contains("Rule violated for bundle $testDirectory.name: $errorMessage")
+
+        where:
+        tasksPaths                                             | reportTaskName                              | limit                                  | errorMessage
+        TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS       | 'jacocoTestCoverageVerification'            | Insufficient.LINE_METRIC_COVERED_RATIO | 'lines covered ratio is 1.0, but expected maximum is 0.5'
+        INTEG_TEST_AND_JACOCO_COVERAGE_VERIFICATION_TASK_PATHS | 'jacocoIntegrationTestCoverageVerification' | Insufficient.CLASS_METRIC_MISSED_COUNT | 'classes missed count is 0.0, but expected minimum is 0.5'
+    }
+
+    def "task is never UP-TO-DATE as it does not define any outputs"() {
```

Comment:
I don't think this is correct.  There *is* an output. It's just not captured as a file anywhere and we don't have a way to capture non-file outputs.  We should make this produce a report file (just a text file) for now so we don't re-run this every time if nothing has changed.